### PR TITLE
Reduce water new branch

### PR DESCRIPTION
**Pull request by MeSophie**
Fixes Partially #8352
@rdstern , @lilyclementsit This PR replace PR #8471  I will close it because of conflicts. @rdstern  can you check if the result producing by End of Rain dialog are correct now?

### CODE REVIEW

**Review by rdstern — CHANGES_REQUESTED** [review]
@MeSophie I think this is getting close.

image

When I return to the dialog as shown, sometimes Ok isn't enabled - though in this example above it should be.  That makes it hard to test.

Note also the fields for the evaporation - both fixed and variable have disappeared.  They don't come back even when pressing reset.

I tested with Dodoma, which had misssing values and that seems ok.  But the transform doesn't work yet with the missing values.

**Comment by MeSophie** [comment]
@lilyclements I hope you can help. Trying to test  ` End of Rain dialog` I checked` variable rdo` using this code

```
# Code generated by the dialog, End of Rains/Season

rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=0, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, ifelse(test=is.na(x=tmin), yes=5, no=tmin), y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
rain_max <- instat_calculation$new(type="calculation", function_exp="ifelse(yes=100, test=is.na(x=rain), no=rain)", result_name="rain_max", calculated_from=list("dodoma"="rain"))
wb_max <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, ifelse(test=is.na(x=tmin), yes=5, no=tmin), y), 0), 100), x=tail(x=rain_max, n=-1), init=0, accumulate=TRUE)", result_name="wb_max", sub_calculations=list(rain_max))
wb <- instat_calculation$new(type="calculation", function_exp="ifelse(test=(wb_min != wb_max) | is.na(x=rain), yes=NA, no=wb_min)", result_name="wb", sub_calculations=list(wb_min, wb_max))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb <= 0.5) | is.na(x=rain)", sub_calculations=list(wb))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season"))
```
I got this error:

image
image
image

I tried also to investigate in R-Studio I optained this:
image

**Comment by Patowhiz** [comment]
@MeSophie what is the status of this PR? Can @rdstern start reviewing?

**Comment by MeSophie** [comment]
> @MeSophie what is the status of this PR? Can @rdstern start reviewing?

@Patowhiz  the dialog is not Working well. It is giving wrong answers so I asked @lilyclements  to check the code. Look like she is busy.

**Comment by Patowhiz** [comment]
@rdstern is this urgent for the next release? If it is, @MeSophie and I can try to access @lilyclements with the specifics required from her.

**Comment by lilyclements** [comment]
@MeSophie looks like you've made some really good progress, thank you for being so patient here. We're nearly _nearly_ there.

I'm comparing the R code output you've given here to what we discussed in #8471

```
# In #8471 you have 
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_min, n = -1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))

# Here you have
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, ifelse(test=is.na(x=tmin), yes=5, no=tmin), y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
```

So in #8471 you have the function:
```
Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_min, n = -1), init=0, accumulate=TRUE)
```

And here you are running
```
Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, ifelse(test=is.na(x=tmin), yes=5, no=tmin), y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)
```

**Conclude:**
Here you want to replace the NAs with 5 in your dialog. You have done that correctly in the `function_exp` of `wb_min`, but you are still using `rain_min`. You can just use `rain` here, and not worry about `rain_min` at all :) 

**To Add**
You can ignore this rain_min and rain_max stuff, surely, and just instead run

```
rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=5, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))

conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb_min <= 0.5) | is.na(x=rain)", sub_calculations=list(wb_min))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb_min)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season"))
```

**Comment by lilyclements** [comment]
@MeSophie looks like you've made some really good progress, thank you for being so patient here. We're nearly _nearly_ there.

@MeSophie Can you point me to where we have said to add this "Replace NA with" checkbox in?

@rdstern Do we want the option to replace missing values with a number to be a checkbox, or is it always there?
If it is always there, why do we have wb_min and wb_max?

Back to the problem

We usually do `rain_min` and `rain_max` to deal with the missing values. However, here you have replaced the missing values with a `5`.
I would expect that that `replace missing values` option is a checkbox.
When we do replace the missing values, we don't need rain_min, rain_max, wb_min, wb_max, we just set all NAs to be a 5 and go from there - 

```
rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=5, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb_min <= 0.5) | is.na(x=rain)", sub_calculations=list(wb_min))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb_min)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb_min", "rain_min", "grouping_by_station_year", "doy_filter", "end_season"))
```

@MeSophie does that make sense?

**Comment by lilyclements** [comment]
To add - if we have no missing values then we run:

```
rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=0, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_min, n = -1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
rain_max <- instat_calculation$new(type="calculation", function_exp="ifelse(yes=100, test=is.na(x=rain), no=rain)", result_name="rain_max", calculated_from=list("dodoma"="rain"))
wb_max <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_max, n = -1), init=0, accumulate=TRUE)", result_name="wb_max", sub_calculations=list(rain_max))
wb <- instat_calculation$new(type="calculation", function_exp="ifelse(test=(wb_min != wb_max) | is.na(x=rain), yes=NA, no=wb_min)", result_name="wb", sub_calculations=list(wb_min, wb_max))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb <= 0.5) | is.na(x=rain)", sub_calculations=list(wb))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_season_date <- instat_calculation$new(type="summary", function_exp="dplyr::if_else(condition=is.na(x=dplyr::first(x=wb)), true=as.Date(NA), false=dplyr::first(x=date))", result_name="end_season_date", save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season, end_season_date))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season", "end_season_date"))
```

As discussed in #8471

**Comment by MeSophie** [comment]
@lilyclements Sorry for the late reply. Due to heavy rain, we were without electricity for a few days.

image
The only discussion that I found about the missing value is this on issue #8352
image

**Comment by MeSophie** [comment]
image

@lilyclements Now I'm a bit confused. Do we need to change the value of missing value in `wb_min/wb_max` by `rain_min/rain_max`? If yes why do we need the capacity and the Replace NAs Value control?  Because if the parameter of the missing value is `yes`, for `rain_min `it is always **0**, for `rain_max` it is changing according to the value of `Capacity control.`

**Comment by lilyclements** [comment]
@MeSophie now I’m confused too! Where did this “Replace NA with” label and box come from?
I think when we get to that then I’ll be able to verify the code we want to run.

**Comment by MeSophie** [comment]
> @MeSophie now I’m confused too! Where did this “Replace NA with” label and box come from? I think when we get to that then I’ll be able to verify the code we want to run.

@lilyclements the only place I saw this label is here
image

**Comment by MeSophie** [comment]
@lilyclements I don't know if `yes` parameter has different meaning in the code but if they are all represent the missing value  and we have to replace the NAs with `rain_min/rain_max`, then certain controls seem unnecessary to me. Or these controls are not well linked. Maybe I'm wrong but using the dialog I notice that the value of `yes` in `rain_min` is always 0 and for `rain_max `is taking from capacity control.

**Comment by lilyclements** [comment]
@MeSophie ok, I think we're both confused. It's probably coming from me so sorry.

>I notice that the value of yes in rain_min is always 0 and for rain_max is taking from capacity control.

My understanding is that the NAs (`yes` value) should be replaced with 0 for `rain_min` and with capacity for `rain_max`. @rdstern can you confirm this?

If this is true, do we need this "Replace NAs as"?

**Comment by rdstern** [comment]
@MeSophie and @lilyclements I am really sorry this is taking you guys so long.  If I had known I would have simply left it out.  Good lesson for the future perhaps.  
I am happy with your rules for min and max.
And I am not sure where the Replace NA with 5 in the dialog has come from.  Do we need it?

If it continues to complicate our life, then I wonder about only allowing this option when there are no missing values!

**Comment by lilyclements** [comment]
@rdstern @MeSophie thanks for this. I'm not sure where this replace NA with 5 came from either.

@MeSophie ok lets remove "Replace NA with". The code we want to run is

```
rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=0, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_min, n = -1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
rain_max <- instat_calculation$new(type="calculation", function_exp="ifelse(yes=100, test=is.na(x=rain), no=rain)", result_name="rain_max", calculated_from=list("dodoma"="rain"))
wb_max <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x = tail(x = rain_max, n = -1), init=0, accumulate=TRUE)", result_name="wb_max", sub_calculations=list(rain_max))
wb <- instat_calculation$new(type="calculation", function_exp="ifelse(test=(wb_min != wb_max) | is.na(x=rain), yes=NA, no=wb_min)", result_name="wb", sub_calculations=list(wb_min, wb_max))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb <= 0.5) | is.na(x=rain)", sub_calculations=list(wb))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_season_date <- instat_calculation$new(type="summary", function_exp="dplyr::if_else(condition=is.na(x=dplyr::first(x=wb)), true=as.Date(NA), false=dplyr::first(x=date))", result_name="end_season_date", save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season, end_season_date))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season", "end_season_date"))
```

As discussed in #8471

**Comment by MeSophie** [comment]
image

@lilyclements I assume that this code is when the rdo value is checked. My question now is what happen when the rdo variable is checked and also reducing WB box is checked?
image

**Comment by MeSophie** [comment]
@lilyclements 1) Please are we removing the `Ifelse` function on `wb_min` and `wb_max` only when reducing box checked?
2) Also this is the actual code 

```
# Dialog: End of Rains/Season

rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=0, no=rain)", result_name="rain_min", calculated_from=list("dodoma"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, tmax, y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE)", result_name="wb_min", sub_calculations=list(rain_min))
rain_max <- instat_calculation$new(type="calculation", function_exp="ifelse(yes=100, test=is.na(x=rain), no=rain)", result_name="rain_max", calculated_from=list("dodoma"="rain"))
wb_max <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, tmax, y), 0), 100), x=tail(x=rain_max, n=-1), init=0, accumulate=TRUE)", result_name="wb_max", sub_calculations=list(rain_max))
wb <- instat_calculation$new(type="calculation", function_exp="ifelse(test=(wb_min != wb_max) | is.na(x=rain), yes=NA, no=wb_min)", result_name="wb", sub_calculations=list(wb_min, wb_max))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb <= 0.5) | is.na(x=rain)", sub_calculations=list(wb))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("dodoma"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(dodoma="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("dodoma"="doy_366"), save=2)
end_season_date <- instat_calculation$new(type="summary", function_exp="dplyr::if_else(condition=is.na(x=dplyr::first(x=wb)), true=as.Date(NA), false=dplyr::first(x=date))", result_name="end_season_date", save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season, end_season_date))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season", "end_season_date"))
```

And it is giving the following error. I think it is coming from stand alone function. Notice that this the using variable here is `tmax`
image

After I change the variable to `rain` and the code is still produicing an error
image


3) Can I remove `Replace NA With` control since it is not using anywhere?

**Comment by lilyclements** [comment]
@MeSophie 

> are we removing the Ifelse function on wb_min and wb_max only when reducing box checked?

We never want to run have the `ifelse(test=is.na(x=doy_366), yes=5, no=doy_366)`. I'm a bit confused where this came from / why it got added in. This is saying if we have a missing DOY to have a 5, otherwise to have the DOY. I don't know why we would ever want to have this.

> Also this is the actual code

Ok so here have you tried to use `tmax` as your `VARIABLE`? A few things to say, if that is the case

1. Nice that you've got the R code to work! Has this come from the dialog or from you? If from the dialog, nice one - looks like you're there! (If it's not from the dialog, then it's great we're on the same page about the code).
2. The errors are coming from what you've used as your `VARIABLE`. The first error I think is occurring because `tmax` has NAs in it. Your second error is having issues because it's too big. So I assume this is coming from using a variable in there. 
Try creating a dummy variable in your data set containing just `1` or something, and test with that instead? 

> 3. Can I remove Replace NA With control since it is not using anywhere?

Yes that can be removed completely.

**Comment by MeSophie** [comment]
@lilyclements I tried to filter the variable to just one year and also created a dummy variable but I still obtaining error with the code even if I use Tmax, dummy variable or rain the are giving this same error
image

**Comment by lilyclements** [comment]
Great!

1. take that script you’re getting and set tmax or your dummy variable etc to be 5, does it work?

2. If you add in ‘calculated_from’ too as another parameter in your function_exp wherever you have the variable, does it work?
(Like you do for rain_max where you run calculated_from=list("dodoma"="rain")))

**Comment by MeSophie** [comment]
image

@lilyclements I obtained this error
image


image
I obtained this error traying to run wb_min line 2.
image

```
rain_min <- instat_calculation$new(type="calculation", function_exp="ifelse(test=is.na(x=rain), yes=0, no=rain)", result_name="rain_min", calculated_from=list("Sheet1"="rain"))
wb_min <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, rain, y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE, calculated_from=list("Sheet1"="rain"))", result_name="wb_min", sub_calculations=list(rain_min))
rain_max <- instat_calculation$new(type="calculation", function_exp="ifelse(yes=100, test=is.na(x=rain), no=rain)", result_name="rain_max", calculated_from=list("Sheet1"="rain"))
wb_max <- instat_calculation$new(type="calculation", function_exp="Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, rain, y), 0), 100), x=tail(x=rain_max, n=-1), init=0, accumulate=TRUE, calculated_from=list("Sheet1"="rain"))", result_name="wb_max", sub_calculations=list(rain_max))
wb <- instat_calculation$new(type="calculation", function_exp="ifelse(test=(wb_min != wb_max) | is.na(x=rain), yes=NA, no=wb_min)", result_name="wb", sub_calculations=list(wb_min, wb_max))
conditions_filter <- instat_calculation$new(type="filter", function_exp="(wb <= 0.5) | is.na(x=rain)", sub_calculations=list(wb))
grouping_by_station_year <- instat_calculation$new(type="by", calculated_from=list("Sheet1"="year"))
doy_filter <- instat_calculation$new(type="filter", function_exp="doy_366 >= 1 & doy_366 <= 366", calculated_from=calc_from_convert(x=list(Sheet1="doy_366")))
end_season <- instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::first(x=wb)), yes=NA, no=dplyr::first(x=doy_366))", result_name="end_season", calculated_from=list("Sheet1"="doy_366"), save=2)
end_season_date <- instat_calculation$new(type="summary", function_exp="dplyr::if_else(condition=is.na(x=dplyr::first(x=wb)), true=as.Date(NA), false=dplyr::first(x=date))", result_name="end_season_date", save=2)
end_of_season_combined <- instat_calculation$new(type="combination", manipulations=list(conditions_filter, grouping_by_station_year, doy_filter), sub_calculations=list(end_season, end_season_date))
data_book$run_instat_calculation(calc=end_of_season_combined, display=FALSE)
rm(list=c("end_of_season_combined", "conditions_filter", "wb", "wb_min", "rain_min", "wb_max", "rain_max", "grouping_by_station_year", "doy_filter", "end_season", "end_season_date"))
```

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 1]
```diff
@@ -2908,101 +2908,27 @@ get_vignette <- function (package = NULL, lib.loc = NULL, all = TRUE)
...
-  example_text <- readLines(tf)
-  example_text <- paste(example_text, collapse = "\n")
-  if (give.lines) {
-    return(example_text)
-  }
-  if (echo) {
-    cat(example_text)
-  }
-  return(example_text)
-}
```
[the remark this answers is not in the document]
Comment (lilyclements) [reply instat/static/InstatObject/R/stand_alone_functions.R:3008]: how recently have these been added to the standalonefunctions.R file? They look recent so I assume they've been accidentally removed to handle a conflict or something. We definitely do not want to remove these in this PR

**Comment by MeSophie** [comment]
> @MeSophie @lilyclements I noticed in this PR a couple of functions used in different dialogues have been removed i.e `cumulative_inventory` `convert_to_list ` `getExample` `getRowHeadersWithText`. Is there any reason I'm ignoring for removing them?

@N-thony  It was a mistake from my side trying to solve conflicts.

**Comment by MeSophie** [comment]
@lilyclements  I retrying to test everything that a did before . But I'm still having an error. This is displayed if I use rain/tmax/dummy variable. 
image

And also for more than one year I obtain the errror size
image.

image
I also try this and obtained error also no matter the variable I use.
image.

image
This way is working, which is perfectly normal because it is the same as using the value and not the variable.

**Comment by MeSophie** [comment]
> @MeSophie can this be done in one week or it still depend on @lilyclements to estimate?

@N-thony  It is still depend on @lilyclements  to estimate because the code is not working.

**Comment by lilyclements** [comment]
@MeSophie you have done a very impressive job here. There is a problem I need to work on on debugging if "Variable" is chosen. Thank you for your initial investigations - it's really useful to know the problem still occurs even if we use a dummy variable which is "5" (but no problem if we enter "5" as the value).

I suggest that you disable the variable option for now so that we just have the "value" option. 

I assume you won't have time to do the following for next week, which is all fine. I can write it up as a PR. But

* To make this work, we want to do something very simple to the code: We want to replace `Reduce` with `accumulate` from the `purrr` package. So can you replace the instances of `Reduce` with `purrr::accumulate`

```
Reduce(f=function(x, y) pmin(pmax(x + y - WB_evaporation(x, 0.5, 100, 5, y), 0), 100), x=tail(x=rain_min, n=-1), init=0, accumulate=TRUE
```

```
purrr::accumulate(.f = ~ pmin(pmax(..1 + ..2 - WB_evaporation(..1, 0.5, 100, 5, ..2), 0), 100), .x=tail(x=rain_min, n=-1), .init=0)
```

* We want to replace `f`, `x`, and `init` with `.f`, `.x`, and `.init`
* We want to remove the argument `accumulate = TRUE` from the `purrr::accumulate` function
* And finally, we need to change `x` and `y` to `..2` and `..1` respectively. 

Then adding a dummy variable will be very simple - something like

```
purrr::accumulate2(.f = ~ pmin(pmax(..1 + ..2 - WB_evaporation(..1, 0.5, 100, ..3, ..2), 0), 100), .x=tail(x=rain_min, n=-1), .y = tail(x=dummy, n=-1), .init=0)
```

@MeSophie if you do not have time then let me know and I'll write this up as a PR :)


_If you were interested:_
The problem is that in this code, with `Reduce` we are iterating through each row of data. 
So, with `Reduce` it takes that current rainfall value from our `x =` argument.
However, we want to take that rainfall value _and_ the evaporation variable from that row. So, we want to change this code entirely to not use `Reduce` and instead we want to use `purrr::accumulate`. 

`Reduce` is what was used before we added in the evaporation changes, and it was good but not the Tidyverse way. `purrr` has more flexibility because we have the option of `accumulate` and `accumulate2`.
I assume this task is too big to be done by next week?

**Comment by MeSophie** [comment]
@lilyclements  Thanks for the update and sorry for my late reply, I didn't mean to say anything untrue. The new function runs without a hitch and the changes are visible in the PR #8698 . I'm closing this because of the conflicts.

**Pull request closed without merging**